macrobot/macrobot/handler.go

```diff
@@ -0,0 +1,176 @@
+package macrobot
+
+import (
+	"database/sql"
+	"strings"
+
+	_ "github.com/go-sql-driver/mysql"
+	"github.com/keybase/go-keybase-chat-bot/kbchat"
+	"github.com/keybase/go-keybase-chat-bot/kbchat/types/chat1"
+	"github.com/keybase/managed-bots/base"
+)
+
+type Handler struct {
+	*base.DebugOutput
+
+	stats *base.StatsRegistry
+	kbc   *kbchat.API
+	db    *DB
+}
+
+var _ base.Handler = (*Handler)(nil)
+
+var whiteListedCommands = [...]string{
+	"flip",
+	"giphy",
+}
+
+func NewHandler(stats *base.StatsRegistry, kbc *kbchat.API, debugConfig *base.ChatDebugOutputConfig, db *DB) *Handler {
+	return &Handler{
+		DebugOutput: base.NewDebugOutput("Handler", debugConfig),
+		stats:       stats.SetPrefix("Handler"),
+		kbc:         kbc,
+		db:          db,
+	}
+}
+
+func (h *Handler) HandleNewConv(conv chat1.ConvSummary) error {
+	welcomeMsg := "I can create and run simple macros! Try `!macro create` to get started."
+	return base.HandleNewTeam(h.stats, h.DebugOutput, h.kbc, conv, welcomeMsg)
+}
+
+func (h *Handler) HandleCommand(msg chat1.MsgSummary) error {
+	if msg.Content.Text == nil {
+		return nil
+	}
+
+	cmd := strings.TrimSpace(msg.Content.Text.Body)
+
+	if !strings.HasPrefix(cmd, "!macro") {
+		return nil
+	}
+
+	tokens, userErr, err := base.SplitTokens(cmd)
+	if err != nil {
+		return err
+	} else if userErr != "" {
+		h.ChatEcho(msg.ConvID, userErr)
+		return nil
+	}
+
+	switch {
+	case strings.HasPrefix(cmd, "!macro run"):
+		return h.handleRun(msg, tokens[2:])
+	case strings.HasPrefix(cmd, "!macro create"):
+		return h.handleCreate(msg, tokens[2:])
+	case strings.HasPrefix(cmd, "!macro list"):
+		return h.handleList(msg)
+	case strings.HasPrefix(cmd, "!macro remove"):
+		return h.handleRemove(msg, tokens[2:])
+	default:
+		h.ChatEcho(msg.ConvID, "Unknown command %q", cmd)
+		return nil
+	}
+}
+
+func (h *Handler) handleRun(msg chat1.MsgSummary, args []string) error {
+	if len(args) != 1 {
+		h.ChatEcho(msg.ConvID, "Invalid number of arguments. Expected one: <name>")
+		return nil
+	}
+
+	macroName := args[0]
+	macroMessage, err := h.db.Get(msg.Channel, macroName)
+	switch err {
+	case nil:
+	case sql.ErrNoRows:
+		h.ChatEcho(msg.ConvID, "Macro '%s' is not defined for this %s", macroName, getChannelType(msg.Channel))
```

Comment:
`%q`

Reply:
How come %q instead of %s? My understanding is it will also escape the string, do we want that?

Reply:
seems fine